traffic_ops/client_tests/mock.go

```diff
@@ -0,0 +1,247 @@
+	helpLong := flag.Bool("help", false, "Print usage information and exit")
+	helpShort := flag.Bool("h", false, "Print usage information and exit")
+	keyPathLong := flag.String("key-path", "", "Specify a path to an SSL private key for the mock server to use (default ./localhost.key)")
+	keyPathShort := flag.String("k", "", "Specify a path to an SSL private key for the mock server to use (default ./localhost.key)")
+	listenAddrLong := flag.String("listen", "", "Choose the address or hostname on which the mock server will listen (default '')")
+	listenAddrShort := flag.String("l", "", "Choose the address or hostname on which the mock server will listen (default '')")
+	portLong := flag.Uint("port", 443, "Choose the port on which the mock server will listen for incoming connections")
+	portShort := flag.Uint("p", 443, "Choose the port on which the mock server will listen for incoming connections")
+	versionLong := flag.Bool("version", false, "Print version information and exit")
+	versionShort := flag.Bool("v", false, "Print version information and exit")
+
+	flag.Parse()
+
+	version := *versionLong || *versionShort
+	if version {
+		fmt.Printf("Traffic Ops Mock server, built for version %s\n", VERSION)
+		os.Exit(0)
+	}
+
+	help := *helpLong || *helpShort
+	if help {
+		flag.Usage()
+		os.Exit(0)
+	}
+
+	var port uint
+	if *portLong != 443 {
+		if *portLong == 0 || *portLong >= 65536 {
+			log.Fatalf("Invalid port number: %d\n", *portLong)
+		}
+		if *portShort != 443 && *portShort != *portLong {
+			log.Fatalf("Must specify exactly one port number!\n")
+		}
+		port = *portLong
+	} else if *portShort != 443 {
+		if *portShort == 0 || *portShort >= 65536 {
+			log.Fatalf("Invalid port number: %d\n", *portLong)
+		}
+		port = *portShort
+	} else {
+		port = 443
+	}
+
+	listenAddr := ""
+	if *listenAddrLong != "" {
+		if *listenAddrShort != "" && *listenAddrShort != *listenAddrLong {
+			log.Fatalf("Must specify exactly one listen address!\n")
+		}
+		listenAddr = *listenAddrLong
+	} else if *listenAddrShort != "" {
+		listenAddr = *listenAddrShort
+	}
+
+	certpath := "./localhost.crt"
+	if *certPathLong != "" {
+		if *certPathShort != "" && *certPathShort != *certPathLong {
+			log.Fatalf("Must specify exactly one path to an SSL certificate!\n")
+		}
+		certpath = *certPathLong
+	} else if *certPathShort != "" {
+		certpath = *certPathShort
+	}
+
+	keypath := "./localhost.key"
+	if *keyPathLong != "" {
+		if *keyPathShort != "" && *keyPathShort != *keyPathLong {
+			log.Fatalf("Must specify exactly one path to an SSL private key!\n")
+		}
+		keypath = *keyPathLong
+	} else if *keyPathShort != "" {
+		keypath = *keyPathShort
+	}
+
+	server := &http.Server{
+		Addr:      fmt.Sprintf("%s:%d", listenAddr, port),
+		TLSConfig: &tls.Config{InsecureSkipVerify: false},
+	}
+
+	for i := API_MIN_MINOR_VERSION; i <= API_MAX_MINOR_VERSION; i += 1 {
+		v := "1." + strconv.Itoa(i)
+		log.Printf("Loading API v%s\n", v)
+		http.HandleFunc("/api/"+v+"/ping", ping)
```

Comment:
I was also hoping this set of routes could be taken from the traffic_ops_golang code if possible, so that new routes that get added would also be automatically picked up by Mock TO. That's probably also another fairly difficult thing to wire into this, but I think it would also be pretty valuable.

Reply:
I have a branch somewhere where I was working on being able to do that kind of reflection - at least pulling the names of the routes if not their associated handlers. It's part of a PR for adding an `OPTIONS` method handler to `/api/` itself, but I don't believe the reviewer and I could ever agree on the appropriate response format.

Reply:
Actually wait, part of the prerequisite work for that change has been merged, so I actually think that might be possible.

The next question, though, is how to implement that. Specific routes need specific handlers, so unless we want to implement another CRUDer (to which I vehemently object, because we have a huge amount of endpoints that don't use it) idk how much use that's gonna be.

Reply:
Automatically picking up routes from traffic_ops_golang wouldn't really make sense for the "static response" use case, since each "static response" would require its own handler. If there was a single "generic" handler function that was capable of just taking any struct, using reflection to dynamically populate all of the struct fields, and marshalling it into json for the response, then maybe we would just have to include an instance of the struct type in the `Route` structs that traffic_ops_golang populates in the `Routes` function.